File: frontend/src/lib/oaf-types.ts
export interface OAFConfig {
  name: string;
  vendorKey: string;
  agentKey: string;
  version: string;
  slug: string;

  description: string;
  author: string;
  license: string;
  tags: string[];

  skills?: OAFSkill[];
  packs?: OAFPack[];
  weblets?: OAFWeblet[];
  mcpServers?: OAFMCPServer[];
  agents?: OAFSubAgent[];

  orchestration?: OAFOrchestration;

  tools?: string[];

  model?: OAFModel | string;
  config?: OAFRuntimeConfig;
  memory?: OAFMemory;

  harnessConfig?: Record<string, unknown>;

  instructions?: string;
}

export interface OAFSkill {
  name: string;
  source: string;
  version: string;
  required: boolean;
}

export interface OAFPack {
  vendor: string;
  pack: string;
  version: string;
  required: boolean;
}

export interface OAFWeblet {
  vendor: string;
  weblet: string;
  version: string;
  launch: 'onDemand' | 'background' | 'foreground';
}

export interface OAFMCPServer {
  vendor: string;
  server: string;
  version: string;
  configDir: string;
  required: boolean;
}

export interface OAFSubAgent {
  vendor: string;
  agent: string;
  version: string;
  role: string;
  delegations?: string[];
  required: boolean;
}

export interface OAFOrchestration {
  entrypoint: string;
  fallback?: string;
  triggers?: OAFTrigger[];
}

export interface OAFTrigger {
  event: string;
  action: string;
}

export interface OAFModel {
  provider: string;
  name: string;
  embedding?: string;
}

export interface OAFRuntimeConfig {
  temperature?: number;
  max_tokens?: number;
  require_confirmation?: boolean;
}

export interface OAFMemory {
  type: 'editable' | 'read-only';
  blocks?: Record<string, string>;
}

export const DEFAULT_OAF: OAFConfig = {
  name: '',
  vendorKey: '',
  agentKey: '',
  version: '1.0.0',
  slug: '',
  description: '',
  author: '',
  license: 'MIT',
  tags: [],
  skills: [],
  mcpServers: [],
  agents: [],
  tools: ['Read', 'Edit', 'Bash', 'Glob', 'Grep'],
  config: {
    temperature: 0.7,
    max_tokens: 4096,
  },
  instructions: '',
};

export const AVAILABLE_TOOLS = [
  { name: 'Read', desc: '读取文件内容' },
  { name: 'Edit', desc: '编辑文件（查找替换）' },
  { name: 'Bash', desc: '执行 Shell 命令' },
  { name: 'Glob', desc: '按模式匹配文件' },
  { name: 'Grep', desc: '搜索文件内容' },
];

export const MODEL_ALIASES = ['sonnet', 'opus', 'haiku'] as const;
export type ModelAlias = typeof MODEL_ALIASES[number];

File: agent-framework-deepagents/server/models/oaf_types.py
from typing import Optional, Any
from pydantic import BaseModel, Field


class SkillConfig(BaseModel):
    name: str
    source: str = "local"
    version: str = "1.0.0"
    required: bool = False
    description: str = ""
    allowed_tools: list[str] = Field(default_factory=lambda: ["bash", "python"])


class MCPServerConfig(BaseModel):
    vendor: str = ""
    server: str = ""
    version: str = "1.0.0"
    config_dir: str = ""
    required: bool = False


class SubAgentConfig(BaseModel):
    vendor: str = ""
    agent: str = ""
    version: str = "1.0.0"
    role: str = ""
    delegations: list[str] = Field(default_factory=list)
    required: bool = False
    endpoint: str = ""


class ModelConfig(BaseModel):
    provider: str = "openai"
    name: str = ""
    embedding: str = ""


class RuntimeConfig(BaseModel):
    temperature: float = 0.7
    max_tokens: int = 4096
    require_confirmation: bool = False


class MemoryConfig(BaseModel):
    type: str = "editable"
    blocks: dict[str, str] = Field(default_factory=dict)


class OrchestrationConfig(BaseModel):
    entrypoint: str = "main"
    fallback: str = ""
    triggers: list[dict] = Field(default_factory=list)


class HarnessConfig(BaseModel):
    a2a: dict = Field(default_factory=dict)
    a2ui: dict = Field(default_factory=dict)


class OAFConfig(BaseModel):
    name: str = "Unnamed Agent"
    vendor_key: str = Field(default="local", alias="vendorKey")
    agent_key: str = Field(default="unnamed", alias="agentKey")
    version: str = "1.0.0"
    slug: str = "local/unnamed"
    description: str = ""
    author: str = "@local"
    license: str = "MIT"
    tags: list[str] = Field(default_factory=list)
    system_prompt: str = ""
    skills: list[SkillConfig] = Field(default_factory=list)
    mcp_servers: list[MCPServerConfig] = Field(default_factory=list, alias="mcpServers")
    sub_agents: list[SubAgentConfig] = Field(default_factory=list, alias="agents")
    tools: list[str] = Field(default_factory=list)
    model: ModelConfig = Field(default_factory=ModelConfig)
    runtime_config: RuntimeConfig = Field(default_factory=RuntimeConfig, alias="config")
    memory: MemoryConfig = Field(default_factory=MemoryConfig)
    orchestration: OrchestrationConfig = Field(default_factory=OrchestrationConfig)
    harness_config: HarnessConfig = Field(default_factory=HarnessConfig, alias="harnessConfig")
    raw_frontmatter: dict = Field(default_factory=dict, exclude=True)

    @property
    def has_skills(self) -> bool:
        return len(self.skills) > 0

    @property
    def has_mcp(self) -> bool:
        return len(self.mcp_servers) > 0

    @property
    def has_sub_agents(self) -> bool:
        return len(self.sub_agents) > 0

    @property
    def local_skills(self) -> list[SkillConfig]:
        return [s for s in self.skills if s.source == "local"]

    @property
    def remote_skills(self) -> list[SkillConfig]:
        return [s for s in self.skills if s.source != "local"]

    def get_a2ui_enabled(self) -> bool:
        hc = self.raw_frontmatter.get("harnessConfig", {})
        da = hc.get("deep-agents", {})
        a2ui = da.get("a2ui", {})
        return a2ui.get("enabled", True)

    def get_catalog_id(self) -> str:
        hc = self.raw_frontmatter.get("harnessConfig", {})
        da = hc.get("deep-agents", {})
        a2ui = da.get("a2ui", {})
        return a2ui.get("catalog_id", "https://a2ui.org/specification/v0_8/standard_catalog_definition.json")
